src/System.Collections/tests/Generic/Dictionary/Dictionary.Generic.Tests.ConcurrentAccessDetection.netcoreapp.cs

```diff
@@ -0,0 +1,181 @@
+                //Assert.Equal("ThrowInvalidOperationException_ConcurrentOperationsNotSupported", Assert.Throws<InvalidOperationException>(() => dic.Remove(new DummyRefType() { Value = 1 }, out DummyRefType value)).TargetSite.Name);
+            });
+            customThread.IsBackground = true;
+            customThread.Start();
+
+            //Wait max 5 seconds, could loop forever
+            Assert.True(customThread.Join(TimeSpan.FromSeconds(5)));
+        }
+
+        [Fact]
+        public static void Add_DictionaryConcurrentAccessDetection_Comparer_ReferenceTypeKey()
+        {
+            Thread customThread = new Thread(() =>
+             {
+                 Dictionary<DummyRefType, DummyRefType> dic = new Dictionary<DummyRefType, DummyRefType>(new CustomEqualityComparerDummyRefType());
+                 dic.Add(new DummyRefType() { Value = 1 }, new DummyRefType() { Value = 1 });
+
+                 //break internal state
+                 var entriesType = dic.GetType().GetField("_entries", BindingFlags.NonPublic | BindingFlags.Instance);
+                 var entriesInstance = (Array)entriesType.GetValue(dic);
+                 var field = entriesInstance.GetType().GetElementType();
+                 var entryArray = (Array)Activator.CreateInstance(entriesInstance.GetType(), new object[] { dic.Count });
+                 var entry = Activator.CreateInstance(field);
+                 entriesType.SetValue(dic, entryArray);
+
+                 Assert.NotNull(dic.GetType().GetField("_comparer", BindingFlags.NonPublic | BindingFlags.Instance).GetValue(dic));
+                 Assert.False(dic.GetType().GetGenericArguments()[0].IsValueType);
+                 Assert.Equal("ThrowInvalidOperationException_ConcurrentOperationsNotSupported", Assert.Throws<InvalidOperationException>(() => dic.Add(new DummyRefType() { Value = 1 }, new DummyRefType() { Value = 1 })).TargetSite.Name);
+                 Assert.Equal("ThrowInvalidOperationException_ConcurrentOperationsNotSupported", Assert.Throws<InvalidOperationException>(() => dic[new DummyRefType() { Value = 1 }]).TargetSite.Name);
+                 //Remove is not resilient yet
+                 //Assert.Equal("ThrowInvalidOperationException_ConcurrentOperationsNotSupported", Assert.Throws<InvalidOperationException>(() => dic.Remove(new DummyRefType() { Value = 1 })).TargetSite.Name);
+                 //Assert.Equal("ThrowInvalidOperationException_ConcurrentOperationsNotSupported", Assert.Throws<InvalidOperationException>(() => dic.Remove(new DummyRefType() { Value = 1 }, out DummyRefType value)).TargetSite.Name);
+             });
+            customThread.IsBackground = true;
+            customThread.Start();
+
+            //Wait max 5 seconds, could loop forever
+            Assert.True(customThread.Join(TimeSpan.FromSeconds(5)));
```

Comment:
I would make these timeouts large (eg 30 or 60 sec). Because if the machine is overloaded, they may fail in the passing case - which is bad. I have seen test machines occasionally be very slow. It is OK if they are large, because unless we regress, they will never hit the timeout. All we care about is the entire test run isn't brought down in that case. Each test library gets 20 minutes total.

Reply:
ok